### PR TITLE
doc: Rework Development Process section

### PR DESCRIPTION
**Pull request by nashif**
- Added various contents from stale wiki pages
- Added API life cycle document
- Add pictures showing process in general
- Split the document into multiple sections.

### CODE REVIEW

File: doc/development_process/code_flow.rst
Comment (aescolar) [inline doc/development_process/code_flow.rst]: I know this is a copy paste, but maybe you want to rename it pull request

File: doc/development_process/documentation.rst
Comment (aescolar) [inline doc/development_process/documentation.rst]: typo: provides an-d- input

File: doc/development_process/documentation.rst
Comment (aescolar) [inline doc/development_process/documentation.rst]: typo: functionality the+y+ validate creating

**Review by aescolar — COMMENTED** [review]
Thanks Anas,
apart from the general questions to the APIs part, just a few minor things

File: doc/development_process/api_lifecycle.rst
Comment (nashif) [reply doc/development_process/api_lifecycle.rst]: This is for public APIs of any kind, will have a section about the scope later.

**Comment by nashif** [comment]
whoever has been reviewing this for the last 30 minutes, mercy.

File: doc/development_process/review_process.rst
Comment (dbkinder) [inline doc/development_process/review_process.rst]: Missing a diagram here...

File: doc/development_process/review_process.rst
Comment (dbkinder) [inline doc/development_process/review_process.rst]:
```suggestion
on the fix, severity, and availability of someone to review them (other than the
```

File: doc/development_process/review_process.rst
Comment (dbkinder) [inline doc/development_process/review_process.rst]:
```suggestion
Trivial changes are those that appear obvious enough and do not require maintainer or code-owner
```

File: doc/development_process/review_process.rst
Comment (dbkinder) [inline doc/development_process/review_process.rst]:
```suggestion
All pull requests need to be reviewed and should not be merged by the author
```

File: doc/development_process/review_process.rst
Comment (dbkinder) [inline doc/development_process/review_process.rst]:
```suggestion
- Hot fixes: Fixing CI issues, reverts, and system breakage
```

File: doc/development_process/review_process.rst
Comment (dbkinder) [inline doc/development_process/review_process.rst]: The audit folks would likely take exception to rushed merges without sufficient reviews...

File: doc/development_process/review_process.rst
Comment (dbkinder) [inline doc/development_process/review_process.rst]:
```suggestion
  They are items that we need to execute and drive to closure.
  Use the mailing lists for discussions.
```

File: doc/development_process/review_process.rst
Comment (dbkinder) [inline doc/development_process/review_process.rst]:
```suggestion
Any failures found during the CI test run will result in a negative review
assigned automatically by the CI system.
```

File: doc/development_process/review_process.rst
Comment (dbkinder) [inline doc/development_process/review_process.rst]: also ``- doc build: verify doc generation is successful``

**Review by dbkinder — CHANGES_REQUESTED** [review]
Nice addition to our processing docs...

File: doc/development_process/api_lifecycle.rst
Comment (carlescufi) [inline doc/development_process/api_lifecycle.rst]: `public header file`

File: doc/development_process/api_lifecycle.rst
Comment (carlescufi) [inline doc/development_process/api_lifecycle.rst]: `usage`

File: doc/development_process/api_lifecycle.rst
Comment (carlescufi) [inline doc/development_process/api_lifecycle.rst]: Maybe add a comment saying that there can be additions to a stable API as long as they don't break compatibility?

File: doc/development_process/api_lifecycle.rst
Comment (carlescufi) [inline doc/development_process/api_lifecycle.rst]: - Code using the deprecated API needs to be modified to remove usage of said API

File: doc/development_process/code_flow.rst
Comment (carlescufi) [inline doc/development_process/code_flow.rst]: Zephyr consists of 3 types of branches:

- `master` which contains the latest state of development
- `topic-` topic branches that are used for shared development of a new feature
- `vx.y-branch` branches which track maintenance releases based on a major release

File: doc/development_process/code_flow.rst
Comment (carlescufi) [inline doc/development_process/code_flow.rst]: Do we need this now? we no longer have subsystem branches

File: doc/development_process/code_flow.rst
Comment (carlescufi) [inline doc/development_process/code_flow.rst]: s/development/topic ?

File: doc/development_process/proposals.rst
Comment (carlescufi) [inline doc/development_process/proposals.rst]: Not sure about this. I think we should keep RFCs in either GitHub issues or Pull Requests, not documents?

File: doc/development_process/proposals.rst
Comment (carlescufi) [inline doc/development_process/proposals.rst]: Or Pull Request?

File: doc/development_process/release_process.rst
Comment (carlescufi) [inline doc/development_process/release_process.rst]: s/gatekeeper/release master ?

Comment (carlescufi) [reply doc/development_process/release_process.rst]: Also below

File: doc/development_process/release_process.rst
Comment (carlescufi) [inline doc/development_process/release_process.rst]: topic branches?

File: doc/development_process/release_process.rst
Comment (carlescufi) [inline doc/development_process/release_process.rst]: Add a script to `scripts/` perhaps?

File: doc/development_process/review_process.rst
Comment (carlescufi) [inline doc/development_process/review_process.rst]: s/GIT/Git

File: doc/development_process/review_process.rst
Comment (carlescufi) [inline doc/development_process/review_process.rst]: :file:`CODEOWNERS`

**Review by carlescufi — CHANGES_REQUESTED** [review]
Nice! This is great work Anas. Just some minor comments on my side.

File: doc/development_process/documentation.rst
Comment (nashif) [reply doc/development_process/documentation.rst]: that is in doxygen, unrelated to the above.

File: doc/development_process/proposals.rst
Comment (aescolar) [inline doc/development_process/proposals.rst]: Side comment: (I know this text is mostly a copy but) I think it would be good to eventually add a paragraph in line with what you mentioned in last TSC: that any mayor change or mayor addition should be brought up to the TSC. So the TSC is informed, can provide guidance, and the change, if accepted, is accounted for in the roadmap.

File: doc/development_process/api_lifecycle.rst
Comment (mbolivar) [inline doc/development_process/api_lifecycle.rst]: s/the Zephyr's/Zephyr's/

File: doc/development_process/api_lifecycle.rst
Comment (mbolivar) [inline doc/development_process/api_lifecycle.rst]: s/diferrent/different/

File: doc/development_process/api_lifecycle.rst
Comment (mbolivar) [inline doc/development_process/api_lifecycle.rst]: s/documentated/documented/

File: doc/development_process/api_lifecycle.rst
Comment (mbolivar) [inline doc/development_process/api_lifecycle.rst]: s|Github issues/pull-requests|GitHub issues and pull requests|

File: doc/development_process/code_flow.rst
Comment (mbolivar) [inline doc/development_process/code_flow.rst]: How about: "The zephyr Git repository has three types of branches:"

File: doc/development_process/documentation.rst
Comment (mbolivar) [inline doc/development_process/documentation.rst]: Maybe add:

`In particular, doxygen's XML output is used as an input when producing the Zephyr project's online documentation.`

File: doc/development_process/documentation.rst
Comment (mbolivar) [inline doc/development_process/documentation.rst]: The Zephyr project uses several test methodologies, the most common being the Ztest framework.

(with link to Ztest here, and no `:ref:` below)

File: doc/development_process/documentation.rst
Comment (mbolivar) [inline doc/development_process/documentation.rst]: "should only be done as doxygen comments in specially designated doxygen groups"?

File: doc/development_process/documentation.rst
Comment (mbolivar) [inline doc/development_process/documentation.rst]: "trace them to requirements"?

File: doc/development_process/documentation.rst
Comment (mbolivar) [inline doc/development_process/documentation.rst]: needs

File: doc/development_process/documentation.rst
Comment (mbolivar) [inline doc/development_process/documentation.rst]: identify it

File: doc/development_process/documentation.rst
Comment (mbolivar) [inline doc/development_process/documentation.rst]: Perhaps:

```
The doxygen ``@see`` directive is used to link features and APIs under test, like so::
```

Note you need double backticks in RST for inline code, not just one.

File: doc/development_process/review_process.rst
Comment (mbolivar) [inline doc/development_process/review_process.rst]: missing "q" in file name

**Review by mbolivar — COMMENTED** [review]
Thanks for this! It answered some concerns of mine (specifically around API acceptance into mainline before multiple implementations were merged).

A bunch of random comments about typos. Feel free to ignore anything that was already here (I see there's a lot of copy/pasting going on).

File: doc/development_process/documentation.rst
Comment (nashif) [reply doc/development_process/documentation.rst]: hmm, not the same thing. The idea is to document the entry test functions and not every function available in the test code.

File: doc/development_process/review_process.rst
Comment (aescolar) [inline doc/development_process/review_process.rst]: Some thoughts here:
As it is today, quite a few issues, are indeed discussions. And (when not tagged as bugs) they do provide some benefits (and drawbacks)
"Github issues" do fulfill the role of a discussion medium a bit better than the maillists (less spam, can fix comments, others can follow or ignore at will, the whole thread is keep in one place..).
Though they have the (very) severe issue of their existence not being broadcasted in any way to possible interested parties. Which requires people to constantly monitor the incoming issues.

So maybe we could improve it by being a bit more consistent with issue labels, and have a rule that new discussion issues need to be accompanied with an email to the mail-list ?

File: doc/development_process/review_process.rst
Comment (aescolar) [inline doc/development_process/review_process.rst]: Silly thing: the build is not subject to the HW being available.

File: doc/development_process/review_process.rst
Comment (aescolar) [inline doc/development_process/review_process.rst]: Also BLE tests in BabbleSim

File: doc/development_process/review_process.rst
Comment (aescolar) [inline doc/development_process/review_process.rst]: shouldn't these 2 be part of the "run sanitycheck script" above

File: doc/development_process/review_process.rst
Comment (nashif) [reply doc/development_process/review_process.rst]: not going into the details here. The above is part of sanity testing

File: doc/development_process/review_process.rst
Comment (nashif) [reply doc/development_process/review_process.rst]: good catch

File: doc/development_process/review_process.rst
Comment (nashif) [reply doc/development_process/review_process.rst]: we do have the right labels in place. This section is just highlighting that issues (bugs, enhancements, etc) need to be driven to a closure and should not be open-ended because of unrelated lengthy discussions or kept open to serve informational purpose.

File: doc/development_process/proposals.rst
Comment (nashif) [reply doc/development_process/proposals.rst]: yes, will add.

File: doc/development_process/release_process.rst
Comment (dbkinder) [inline doc/development_process/release_process.rst]: If you're still editing, you should break this row into multiple lines (as you did for the last row in this table)

**Review by carlescufi — APPROVED** [review]

**Review by dbkinder — APPROVED** [review]
looks OK, thanks.

**Pull request merged**
